fsw/src/crypto.c

```diff
@@ -2647,46 +2648,252 @@ static int32 Crypto_PDU(char* ingest,TC_t* tc_frame)
+    temp_tc_header.bypass = ((uint8)in_frame[0] & 0x20) >> 5;
+    temp_tc_header.cc     = ((uint8)in_frame[0] & 0x10) >> 4;
+    temp_tc_header.spare  = ((uint8)in_frame[0] & 0x0C) >> 2;
+    temp_tc_header.scid   = ((uint8)in_frame[0] & 0x03) << 8;
+    temp_tc_header.scid   = temp_tc_header.scid | (uint8)in_frame[1];
+    temp_tc_header.vcid   = ((uint8)in_frame[2] & 0xFC) >> 2;
+    temp_tc_header.fl     = ((uint8)in_frame[2] & 0x03) << 8;
+    temp_tc_header.fl     = temp_tc_header.fl | (uint8)in_frame[3];
+    temp_tc_header.fsn	  = (uint8)in_frame[4];
+
+    // Check if command frame flag set
+    if (temp_tc_header.cc == 1)
+    {
+        /*
+        ** CCSDS 232.0-B-3
+        ** Section 6.3.1
+        ** "Type-C frames do not have the Security Header and Security Trailer."
+        */
+        #ifdef TC_DEBUG
+            OS_printf(KYEL "DEBUG - Received Control/Command frame - nothing to do.\n" RESET);
+        #endif
     }
 
-    //Create Frame Body
-    CFE_PSP_MemCpy(&tempTC[security_header_bytes],&tc_ingest[0],*len_ingest);
-    count+=*len_ingest;
+    // TODO: If command frame flag not set, need to know SDLS parameters
+    // What is the best way to do this - copy in the entire SA for a channel?
+    // Expect these calls will return an error is the SA is not usable
+    // Will need to know lengths of various parameters from the SA DB
+    else
+    {
+        // Query SA DB for active SA / SDLS paremeters
+        // TODO: Likely API call
+        // TODO: Magic to make sure we're getting the correct SA.. 
+        // currently SA DB allows for more than one
+        SecurityAssociation_t temp_SA;
+        int16_t spi = -1;
+        for (int i=0; i < NUM_SA;i++)
+        {
+            if (sa[i].sa_state == SA_OPERATIONAL)
+            {
+                spi = i;
+                //break; // Needs modified to ACTUALLY get the correct SA
+                         // TODO: Likely API discussion
+            }
+        }
 
-    //Create Security Trailer
-    for(;count < tc_size;count++){
-        tempTC[count]=0x55; //put dummy filler bits in security trailer for now.
-    }
+        // If no active SA found
+        if (spi == -1)
+        {
+            OS_printf(KRED "Error: No operational SA found! \n" RESET);
+            status = OS_ERROR;
+        }
+        // Found an active SA
+        else
+        {
+            CFE_PSP_MemCpy(&temp_SA, &(sa[spi]), SA_SIZE);
+            #ifdef TC_DEBUG
+                OS_printf("Operational SA found at index %d.\n", spi);
+            #endif
+        }
+
+        #ifdef SA_DEBUG
+            OS_printf(KYEL "DEBUG - Received data frame.\n" RESET);
+            OS_printf("Using SA 0x%04X settings.\n", spi);
+        #endif
+
+        // Verify SCID 
+        if ((temp_tc_header.scid != temp_SA.gvcid_tc_blk[temp_tc_header.vcid].scid) \
+            && (status == OS_SUCCESS))
+        {
+            OS_printf(KRED "Error: SCID incorrect! \n" RESET);
```

Comment:
This is an unnecessary check. Incorrect SCID would result in no SAs being returned.

Reply:
Should potentially leave some of these for now until SA things are straightened out as they may affect functionality / expected results for some initial test cases as sanity checks.